### DISCUSSION TITLE
Panda + Chakra

### DISCUSSION THREAD

**Discussion by barbinbrad** [chakra-ui/chakra-ui#7785] [General]
Hi guys, thanks so much for all the work, and congratulations on all the great new stuff.

I've recently read [about the future of Chakra](https://www.adebayosegun.com/blog/the-future-of-chakra-ui), and wanted to ask about the possibility of refactoring Chakra to use Panda under the hood?

Currently I use Chakra and shadcn. Both have very nice APIs and DX. Like almost everyone, given the choice, I'd rather not take a performance hit at runtime, which is why I've started moving toward shadcn. You get a similar (way less) component API and DX without the runtime cost. 

In my very limited understanding, Ark is like Radix and Panda is like Tailwind (but better). The reason (I think) that a lot of people love shadcn is that you don't have to know much about Radix or Tailwind to get a great result. The complexity is largely hidden.

Along those lines, I think it's reasonable to think that some smart guy would create a component library that hides the complexity of Ark and Panda. And I guess my question is, why shouldn't that be Chakra? I'm sure it's a massive lift. I'm not sure that it's even possible, but I'm wondering why not?

**Comment by barbinbrad** [comment]
It may be that this _is_ the plan, but based on the this section in the article, I wasn't clear.

> What should I use in my project?
> - If you're okay with runtime CSS-in-JS, use Chakra UI (It comes with Ark and our current CSS-in-JS solution) 
> - If you're not okay with runtime CSS-in-JS, use Ark and Panda

**Reply by coopbri** [reply]
The wording is still vague, but [this tweet](https://twitter.com/thesegunadebayo/status/1643910659141251072) from Segun seems to indicate that Chakra will remain using Ark + Emotion. I do hope I am wrong and that Chakra will migrate to Ark + Panda however (if even possible as you mentioned).

Worth looking at the talk with Lee Robinson a couple years ago as well: https://www.youtube.com/watch?v=I5xEc9t-HZg (~27:55min mark). Seems to indicate Chakra will use the "build-time CSS engine" (now publicly known as Panda), though might have just meant ecosystem.

**EDIT:** for good measure, asked about this in Discord [here](https://discord.com/channels/660863154703695893/1122256822099783763/1122256822099783763).

**Comment by godfrednanaowusu** [comment]
![image_2023-06-26_214714430](https://github.com/chakra-ui/chakra-ui/assets/32749058/0af8119e-933b-4534-89e8-a870322b198b)

I have tried asking about this severally on twitter and I have seen other people interested in this answer too, but unfortunately, the messaging is so confusing.

Look here for a very detailed response:
https://twitter.com/codebender828/status/1650712270325026816

however @segunadebayo says Chakra would continue to be powered by Emotion, this creates confusion.

I would appreciate if we could settle this once and for all.

**Reply by barbinbrad** [reply]
Same here. It's all good either way. I'd just like to know so I can plan. Ultimately, I'd like to get off of emotion. If I can do that without much thinking by just upgrading chakra, that's super wonderful. If it involves a big refactor, that's OK too.

**Comment by godfrednanaowusu** [comment]
We are in the same shoe, I also need it to plan, I am hoping for the best.

**Comment by crhistianramirez** [comment]
Everywhere I've read it (articles, twitter) makes it pretty obvious that Chakra and Panda aren't in the road map. If you need something compatible with react server components then the recommendation is to use ark and panda separately.

Ark seems like a nice library for building component libraries but I'm an application developer and that seems like a lot of investment for what I need to use it for. I am in the same camp that would love to see Chakra evolve to use ark and panda under the hood and unfortunately if that doesn't become possible will likely move on to some other component library that accomplishes something similar. I'd rather not though because I really love Chakra

**Reply by barbinbrad** [reply]
Same. I know this isn't any of my business, and that we're very lucky someone like @segunadebayo has done so much great work. But this seems like a big decision to me. And I think that the likely outcome of this is that 1/ people stop using chakra, and 2/ people move to something more user-friendly like shadcn instead of panda + ark. I'd hate to see that because it seems like all the pieces are in place for Chakra to be _the_ component library for the next 5-10 years if it were refactored with Panda under the hood.

**Comment by coopbri** [comment]
Someone asked about this [on LinkedIn](https://www.linkedin.com/feed/update/urn:li:ugcPost:7086817363173466112?commentUrn=urn%3Ali%3Acomment%3A%28ugcPost%3A7086817363173466112%2C7086823704088698881%29&dashCommentUrn=urn%3Ali%3Afsd_comment%3A%287086823704088698881%2Curn%3Ali%3AugcPost%3A7086817363173466112%29) today as well (no response yet). Just trying to aggregate as much context as possible into this discussion.

**Comment by coopbri** [comment]
We now have a clear answer about the future, Segun answered [on LinkedIn](https://www.linkedin.com/feed/update/urn:li:ugcPost:7086817363173466112?commentUrn=urn%3Ali%3Acomment%3A%28ugcPost%3A7086817363173466112%2C7086823704088698881%29&replyUrn=urn%3Ali%3Acomment%3A%28ugcPost%3A7086817363173466112%2C7087920936313593856%29):

> You can think of Chakra V3 as Ark + Emotion. It’ll feature most of the features in Panda but use Emotion.

> Panda + Ark is pretty much the direction we’re heading in the next years. Headless components combined with build-time CSS in JS. You can go in that direction if eliminating runtime overhead is your goal.

**Reply by barbinbrad** [reply]
RIP Chakra?

**Comment by godfrednanaowusu** [comment]
Finally, we got something

**Comment by coopbri** [comment]
In case you guys haven't seen it, @cschroeter (from the Chakra team) is working on a Panda + Ark toolkit called Park UI: https://github.com/cschroeter/park-ui

**Comment by nickzelei** [comment]
Maybe this is being made intentionally unclear, but I think the main question that everyone is trying to figure out is: Will Chakra v4 be Ark + Panda?

Based on the line in the article OP posted above, that isn't going to be the case.

> When Panda is production-ready, we'll recommend switching to Ark and Panda for new projects.

That to me sounds like Chakra-UI is effectively dead for any project that intends to take advantage of server-side components (so anyone using NextJS).

> * If you're okay with runtime CSS-in-JS, use Chakra UI (It comes with Ark and our current CSS-in-JS solution)
> * If you're not okay with runtime CSS-in-JS, use Ark and Panda

Naively, this could mean that Chakra-UI will eventually switch out Emotion with Panda, but it's not clear that will ever happen. Which is totally ok! I think folks just want some closure knowing that if that's the case, they can move on to something new. Otherwise, continue using Chakra and then eventually benefit from the upgrade.

---
After spending more time this morning re-reading the future of chakra article as well as pouring through the Ark/Panda docs - it seems like the way to go forward is to just use both of those libraries separately.
It's more mental overhead because it's two sets of docs, but in the long run, it's making your codebase less dependent on any one specific library.
Ark is intentionally style-agnostic. This makes it easy to switch out Panda/Tailwind/plain-css/whatever you want in the future.
Panda is actually pretty similar to Chakra. it even generates JSX components that look pretty similar to Chakra!

I'm going to continue evaluating these for our apps internally and look into removing Chakra as a dependency. I'm viewing this as merely a transition from a monolith to microservices, but in react-component terms. :)

**Comment by cschroeter** [comment]
I believe it's reasonable to anticipate that Chakra UI v3 will use Ark UI and continue leveraging Emotion as its styling engine. This transition might introduce breaking changes, but with a bit of effort, these can be addressed.

Integrating Ark UI has notable advantages. It will introduce many new components to the Chakra UI ecosystem and significantly streamline the codebase. Additionally, efforts will be made to harmonize the styling solution with that of Panda CSS.

As for a potential v4, it's crucial to manage expectations. While it's technically feasible to have a Chakra UI version that employs Panda as its styling engine, I'm skeptical. Given the intricacies of static CSS generation, it might not offer a straightforward migration path for larger projects to transition to a v4. At least at the time of writing.

---

On a personal note, I've observed that while monolithic UI component libraries—like Material UI, Mantine, and indeed Chakra UI—offer tremendous value at the onset of projects, they can sometimes become limiting as projects evolve.

For instance, challenges may arise when these libraries lack a specific component, or when developers find themselves unable to customize a component as per their requirements due to the restrictive APIs these frameworks offer. This is one reason we see the rising popularity of solutions like https://ui.shadcn.com/. These platforms capitalize on a headless library approach and layer it with a contemporary styling solution—Tailwind, in this case.

Though the components derived from such platforms might initially appear a tad verbose, they do ensure flexibility and adaptability in the long run.

**Reply by barbinbrad** [reply]
Hey man, thanks for all the work you've done. Agree with basically everything you've said here. Also, Park UI seems cool (I guess you're the "smart guy" I was referencing in the OP).

And I'm sure I don't speak for everyone -- but given this information, Chakra is now basically tech debt -- at least from my limited perspective. So now my question is -- how do I refactor? Do I go the shadcn route? Or do I wait for Park UI to become stable? I sort of lean toward shadcn because I've used it -- it's stable -- and it has all those benefits you've mentioned.  Any recommendations?

**Reply by nickzelei** [reply]
After doing a lot more research, I ended up going with shadcn. I evaluated Park and while it seems great, the DX for shad was a bit higher at the current time. I haven't used tailwind yet, but decided to give it a go. It's kind of a lot to take in considering it's learning tailwind, radix, as well as shad's way of doing things, but it only took a few hours to get the hang of it.

Both projects are great and have a lot of promise. Once Park/Panda gets a bit more developed, it will certainly be a force to be reckoned with!

**Reply by cschroeter** [reply]
I wouldn't label Chakra as tech debt. If your project is running smoothly with it, there's no urgent need to switch, especially based on the discussions in this thread. Remember, a v3 with Ark components is on the horizon, and a potential v4 with Panda might follow. Transitioning from v3 to v4 might be as extensive as switching UI frameworks altogether, albeit perhaps slightly less so.

For new projects, always weigh your options. Choose what best aligns with your project's needs and your personal preferences.
If you like having all styles in a single string, go for Tailwind. If you value maintainable styling with style props, Panda CSS would be the choice.

**Reply by strvbkzhn** [reply]
So, should we wait for Chakra V3 this month?

**Reply by vitalijalbu** [reply]
yes, like last year 🤣

**Comment by coopbri** [comment]
Relevant update: https://www.adebayosegun.com/blog/chakra-panda-ark-whats-the-plan

**Reply by crhistianramirez** [reply]
This is great news. I know it's not happening soon but sounds like by v4 we'll have a supercharged chakra ui powered by ark, and panda. Love it

**Reply by barbinbrad** [reply]
Haha! I'm fired up too (although I don't think he's actually saying that, I _want_ to believe it)!
